File: python/compare_cpp_runtime.py
from pathlib import Path

import numpy as np

from discrete_sensor_model import simulate_discrete


def step_input(t: float) -> float:
    return 0.0 if t < 1.0 else 1.0


repo_root = Path(__file__).resolve().parents[1]
csv_file = repo_root / "cpp" / "cpp_runtime.csv"

cpp = np.genfromtxt(
    csv_file,
    delimiter=",",
    names=True,
)

time_py, state_py, _, _ = simulate_discrete(
    k=1.2,
    r=0.2,
    u_fun=step_input,
    sample_time=0.005,
    t_span=(0.0, 12.0),
    x0=0.0,
    y0=0.0,
)

if len(cpp["time"]) != len(time_py):
    raise RuntimeError(
        f"Length mismatch: C++={len(cpp['time'])}, Python={len(time_py)}"
    )

error_t = cpp["time"] - time_py
error_x = cpp["x"] - state_py[0, :]
error_y = cpp["y"] - state_py[1, :]

print("Generated C++ vs Python discrete reference")
print("------------------------------------------")
print(f"Samples          : {len(time_py)}")
print(f"Max time error   : {np.max(np.abs(error_t)):.17g}")
print(f"x max abs error  : {np.max(np.abs(error_x)):.17g}")
print(f"x RMS error      : {np.sqrt(np.mean(error_x**2)):.17g}")
print(f"y max abs error  : {np.max(np.abs(error_y)):.17g}")
print(f"y RMS error      : {np.sqrt(np.mean(error_y**2)):.17g}")

File: matlab/compare_cpp_runtime.m
function report = compare_cpp_runtime(params)
%COMPARE_CPP_RUNTIME Compare generated C++ runtime output with MATLAB discrete reference.

if nargin < 1
    params = init_params();
end

% Repository root:
% .../mems-sensor-simulink/matlab/compare_cpp_runtime.m
%                    ↑ repo root
repo_root = fileparts(fileparts(mfilename('fullpath')));

csv_file = fullfile(repo_root, 'cpp', 'cpp_runtime.csv');

if exist(csv_file, 'file') ~= 2
    error('compare_cpp_runtime:MissingCSV', ...
        'C++ runtime CSV not found: %s', csv_file);
end

%% Load generated C++ runtime output
cpp = readtable(csv_file);

%% Generate MATLAB discrete reference
[t_matlab, state_matlab] = simulate_discrete_model(params);

%% Basic consistency checks
if height(cpp) ~= numel(t_matlab)
    error('compare_cpp_runtime:LengthMismatch', ...
        'C++ has %d samples, MATLAB has %d samples.', ...
        height(cpp), numel(t_matlab));
end

time_error = max(abs(cpp.time - t_matlab));

%% Numerical errors
error_x = cpp.x - state_matlab(:, 1);
error_y = cpp.y - state_matlab(:, 2);

report.max_abs_error_time = time_error;

report.max_abs_error_x = max(abs(error_x));
report.rms_error_x = sqrt(mean(error_x.^2));

report.max_abs_error_y = max(abs(error_y));
report.rms_error_y = sqrt(mean(error_y.^2));

%% Print report
fprintf('\nGenerated C++ vs MATLAB discrete reference\n');
fprintf('------------------------------------------\n');
fprintf('Samples          : %d\n', height(cpp));
fprintf('Max time error   : %.17g s\n', report.max_abs_error_time);
fprintf('x max abs error  : %.17g\n', report.max_abs_error_x);
fprintf('x RMS error      : %.17g\n', report.rms_error_x);
fprintf('y max abs error  : %.17g\n', report.max_abs_error_y);
fprintf('y RMS error      : %.17g\n', report.rms_error_y);

end